reference/programming-model.md

```diff
@@ -189,9 +189,100 @@ url := virtualmachine.DnsName().Apply(func(dnsName string) (interface{}, error)
+  records: [certCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordValue)],
+  ...
+```
+
+```typescript
+let certCertificate = new aws.acm.Certificate("cert", {
+  domainName: "example.com",
+  validationMethod: "DNS",
+});
+let certValidation = new aws.route53.Record("cert_validation", {
+  // Need to pass along a deep subproperty of this Output
+  records: [certCertificate.domainValidationOptions.apply(domainValidationOptions => domainValidationOptions[0].resourceRecordValue)],
+  ...
+```
+
+```python
+certificate = aws.acm.Certificate("cert",
+  domain_name="example.com",
+  validation_method="DNS",
+  
+record = aws.route53.Record("validation",
+  # Need to pass along a deep subproperty of this Output
+  records=[certificate.domain_validation_options.apply(
+      lambda domain_validation_options: domain_validation_options[0].resource_record_value
+  )],
+  ...
+```
+
+```go
+// Helpers for accessing properties are not yet available in Go.
+// 
+// See https://github.com/pulumi/pulumi/issues/1614.
+```
+
+To make this kind of property and array-element access more simple, `Outputs` 'lift' the properties of the values they wrap to themselves as well.  For example, if the wrapped property contains a value `prop: string`, then the `Output` wrapping that has a property `prop: Output<string>`.  This 'lifting' is just shorthand for writing out the full `.apply` call.  In other words, an expression of the form `someOutput.someProp` is equivalent to `someOutput.apply(o => o.someProp)`.  Similarly, `someOutput[val]` is equivalent to writing `someOutput.apply(o => o[val])` (with the right syntax for each respective language).  Given this, the above examples can be rewritten into:
+
+{% include langchoose.html %}
+
+```javascript
+let certCertificate = new aws.acm.Certificate("cert", {
+  domainName: "example.com",
+  validationMethod: "DNS",
+});
+let certValidation = new aws.route53.Record("cert_validation", {
+  records: [certCertificate.domainValidationOptions[0].resourceRecordValue],
+  ...
+```
+
+```typescript
+let certCertificate = new aws.acm.Certificate("cert", {
+  domainName: "example.com",
+  validationMethod: "DNS",
+});
+let certValidation = new aws.route53.Record("cert_validation", {
+  records: [certCertificate.domainValidationOptions[0].resourceRecordValue],
+  ...
+```
+
+```python
+certificate = aws.acm.Certificate("cert",
+  domain_name="example.com",
+  validation_method="DNS",
+  
+record = aws.route53.Record("validation",
+  records=[certificate.domain_validation_options[0].resource_record_value],
+  ...
+```
+
+```go
+// Helpers for accessing properties are not yet available in Go.
+// 
+// See https://github.com/pulumi/pulumi/issues/1614.
+```
+
+This greatly helps the clarity of the final code, while not losing any important dependency information that is needed for properly creating and maintaining the stack.
```

Comment:
I think we can avoid words like “greatly” here.  Also not exactly sure how the dependency information part is related - that’s a given for anything related to Ouput right?  Doesn’t seem that property access changes it?

Might be worth adding note here on calling out the related scenarios which will require using apply (method calls, passing an output to another function, etc. ).

Reply:
> that’s a given for anything related to Ouput right? Doesn’t seem that property access changes it?

I want to help people realize you get an Output back, and it's not like `x.y.z` is just giving a string back that loses it.

Reply:
removed the 'greatly' part.